Friend notifcation improved

ThunderIW opened the pull request:
The system can do the following things

- User can send a friend Request to a particular user by using the addFriendModal which has been modified to be added to newly created fields in the database  so that friends are not forced  to be friends 
- Receiver can accept friend request putting the accepted friend  both the sender and receiver friendlist in the database 
- UI which can be seen below and is integrated into the main appshell 

![image](https://github.com/ColinLefter/Accord/assets/43197270/b3cf0d33-f247-4218-933e-5d1c3335cf4f)


![image](https://github.com/ColinLefter/Accord/assets/43197270/d64851b8-dc18-4686-9f63-f4e4137f9ed7)
 

**receiver end** 
![image](https://github.com/ColinLefter/Accord/assets/43197270/a3a8d4dc-6a57-49af-b4b6-d0d8f9eca0c5)



**sender side**
![image](https://github.com/ColinLefter/Accord/assets/43197270/f525c651-6615-437a-9962-307d4cca28dd)



after accepted friend request both sender and receiver 
![image](https://github.com/ColinLefter/Accord/assets/43197270/49a5bbe7-76bc-49ed-9735-6021c897cd83)
![image](https://github.com/ColinLefter/Accord/assets/43197270/067df602-0e9d-48b9-a91b-e46c20972dad)
new friend thunder90 added to Toby side

TobyNguyen710 commented on Application/Frontend/components/FriendsColumn/AddFriendModal.tsx:
Modified so does not forced adding to the friend list

TobyNguyen710 commented on Application/Frontend/components/FriendsColumn/FriendRequest.jsx:
This is a component for friend request

TobyNguyen710 commented on Application/Frontend/components/Navbar/notifcation.tsx:
This is the notification component

TobyNguyen710 commented on Application/Frontend/package-lock.json:
just a json change

TobyNguyen710 commented on Application/Frontend/package.json:
This is also a json change.

TobyNguyen710 commented on Application/Frontend/pages/TestSendFriendRequestPage.tsx:
This is a test page

TobyNguyen710 commented on Application/Frontend/pages/accord.tsx:
changes in the main accord page

TobyNguyen710 commented on Application/Frontend/pages/api/Get-SentFriendRequest.ts:
Api for get sent friend request

TobyNguyen710 commented on Application/Frontend/pages/api/Get-pendingFriendRequest.ts:
Api for get pending friend request

TobyNguyen710 commented on Application/Frontend/pages/api/accept-friend-request.ts:
api for accept friend request

TobyNguyen710 commented on Application/Frontend/pages/api/add-friend.ts:
modified add friend api

TobyNguyen710 commented on Application/Frontend/pages/api/add_freind_request_to_user.ts:
api to add friend to users

TobyNguyen710 commented on Application/Frontend/pages/api/get-username-from-id.ts:
api to get username from id

TobyNguyen710 commented on Application/Frontend/pages/get_friendRequest.tsx:
Just a test page

TobyNguyen710 commented on Application/Frontend/pages/notify.tsx:
empty file

TobyNguyen710 commented on package-lock.json:
package update

TobyNguyen710 commented on package.json:
pakage update

TobyNguyen710 reviewed (approved):
The entire changes work as intended:
- Able to add friend without forcing to accept friend request
- Able to see friend request notification
- Able to accept the friend request and update the friend list.

ColinLefter commented on Application/Frontend/components/FriendsColumn/AddFriendModal.tsx:
```diff
@@ -1,88 +1,77 @@
-import { useEffect } from 'react';
+import { Modal, Text, Stack, Button, TextInput } from '@mantine/core';
+import { NewFriendModalProps } from '@/accordTypes'; // Make sure this path is correct for your project
+import { showNotification } from '@mantine/notifications';
```
Great use of the built-in notification system from Mantine. Adding responsive notifications throughout the application is one of the next steps toward finalizing our application UX.

ColinLefter commented on Application/Frontend/components/FriendsColumn/AddFriendModal.tsx:
```diff
@@ -1,88 +1,77 @@
-   */
   const handleAddFriendClick = async () => {
-    if (friendUsername) {
+    if (friendUsername.trim()) {
```
This will cause problems if somebody intentionally added whitespace in their username and this type of check is not consistent throughout all other components of the application that rely on username checks. Instead of using `.trim` everywhere, we should ensure users cannot upload usernames with whitespace. However, this would be complicated as the changes in account data are done internally through the provider.

ColinLefter commented on Application/Frontend/components/FriendsColumn/AddFriendModal.tsx:
```diff
@@ -1,88 +1,77 @@
+          showNotification({
+            title: 'Error',
+            message: 'Failed to send friend request.',
+            color: 'red',
```
Great distinction in notification colors. Good UX principles.

ColinLefter commented on Application/Frontend/components/FriendsColumn/AddFriendModal.tsx:
```diff
@@ -1,88 +1,77 @@
-        close();
+        if (searchResult !== null) {
+          setErrorMessage('');
+          setLastFetched(Date.now());
```
Excellent use of our custom context that facilitates real-time updates. Great work in using this!

ColinLefter commented on Application/Frontend/components/FriendsColumn/FriendRequest.jsx:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    try {
+      const response = await fetch('/api/add_freind_request_to_user', {
```
Please update your endpoint names. There is a typo in this endpoint, and it is poorly formatted (please use hyphens instead).

ColinLefter commented on Application/Frontend/components/FriendsColumn/FriendRequest.jsx:
```diff
@@ -0,0 +1,64 @@
+  };
+
+  return (
+    <div>
```
Please use Mantine components. This div can be replaced with a Mantine Stack.

ColinLefter commented on Application/Frontend/components/FriendsColumn/FriendRequest.jsx:
```diff
@@ -0,0 +1,64 @@
+            required
+          />
+        </label>
+        <button type="submit">Send Friend Request</button>
```
Use the button from Mantine. The reason why this is so important is because if I use the theme toggle, none of these HTML elements will be styled accordingly.

ColinLefter commented on Application/Frontend/components/FriendsColumn/FriendRequest.jsx:
```diff
@@ -0,0 +1,64 @@
+        </label>
+        <button type="submit">Send Friend Request</button>
+      </form>
+      {message && <p>{message}</p>}
```
You can condense this to just `message?`

ColinLefter commented on Application/Frontend/components/Navbar/notifcation.tsx:
```diff
@@ -0,0 +1,126 @@
+  const fetchFriendRequests = async () => {
+    setLoading(true);
+    try {
+      const sentResponse = await fetch('/api/Get-SentFriendRequest', {
```
Again, please revise your API endpoint names to follow our naming standard. This endpoint should be renamed to `get-sent-friend-requests`

ColinLefter commented on Application/Frontend/components/Navbar/notifcation.tsx:
```diff
@@ -0,0 +1,126 @@
+        headers: { 'Content-Type': 'application/json' },
+        body: JSON.stringify({ id: userId }),
+      });
+      const receivedResponse = await fetch('/api/Get-pendingFriendRequest', {
```
Rename this API endpoint as well

ColinLefter commented on Application/Frontend/components/Navbar/notifcation.tsx:
```diff
@@ -0,0 +1,126 @@
+  };
+
+  const fetchUsernames = async (friendIds: string[], setFunction: React.Dispatch<React.SetStateAction<FriendRequest[]>>) => {
+    const requests = friendIds.map(id => fetch('/api/get-username-from-id', {
```
This is an example of a well-named API endpoint

ColinLefter commented on Application/Frontend/components/Navbar/notifcation.tsx:
```diff
@@ -0,0 +1,126 @@
+        {sentRequests.length === 0 && !loading && <Text>No sent friend requests</Text>}
+        <Divider />
+        <Menu.Label>Received Friend Requests</Menu.Label>
+        {loading ? <Notification title="Loading..." /> : receivedRequests.map((item) => (
```
Mantine actually had some nice loading skeletons that you could use as well. However, a loading notification is ok for now.

ColinLefter commented on Application/Frontend/pages/api/accept-friend-request.ts:
```diff
@@ -0,0 +1,74 @@
+ *   through environment variables or a configuration file.
+
+ * Usage:
+ * Deploy this function as part of your Next.js API routes. Make sure to handle client-side
```
Remove these GPT comments to improve relevance and readability

ThunderIW replied:
GPT Comments have been removed

ColinLefter commented on Application/Frontend/pages/api/add_freind_request_to_user.ts:
```diff
@@ -0,0 +1,42 @@
+
+      // Assuming `id` is the recipient's user ID. Add `requestingUserId` to the recipient's pendingFriendRequest array.
+      const updateResult = await accountsCollection.updateOne(
+        { _id: new ObjectId(id) }, // Ensure the ID is correctly formatted for MongoDB
```
You should not be using this default is from MongoDB but rather our one. However, if this collection does not have a unique ID field and does not need one, you could keep this as long as it works.

ColinLefter commented on Application/Frontend/pages/api/add_freind_request_to_user.ts:
```diff
@@ -0,0 +1,42 @@
+      client = new MongoClient(getMongoDbUri());
+      await client.connect();
+      const db = client.db('Accord');
+      const accountsCollection = db.collection("Test_friend_list_request_insertion");
```
You also don't need to create a test collection for this particular case as you can use the Accounts collection directly. You can just create two accounts with the provider to test things that require two users. If you are testing fields that don't currently exist in any document in the collection, you can just add them to existing collections as long as they don't break anything or create the new fields as part of account syncing (in the registration endpoint)

ColinLefter reviewed (changes requested):
Everything looks good, but just address the changes I mentioned in the comments. After that, this can be merged safely (assuming no additional bugs are added)

ThunderIW commented:
Fixed all issues, removed test API class, and renamed my APIs that are in use with Hypens 
these included the 
- Get-SentPendingFriendList
- Get-RecivedPendingFriendList 

These are now renamed according 
- Get-Sent-Pending-Friend-List
- Get-Recived-Pending-Friend-List

ThunderIW reviewed (commented):
Fixed all issues, removed test API class,renamed my APIs that are in use with Hypens, remove GPT lines in API
these included the

Get-SentPendingFriendList.ts
Get-RecivedPendingFriendList.ts
These are now renamed according

Get-Sent-Pending-Friend-List.ts
Get-Recived-Pending-Friend-List.ts

ColinLefter reviewed (approved):
Thank you for implementing the requested changes! Great work!
